cmd/otel-allocator/allocation/per_zone.go

```diff
@@ -0,0 +1,185 @@
+	"time"
+)
+
+const perZoneStrategyName = "per-zone"
+const cacheTTL = 120 * time.Minute
+
+type ZonedNode struct {
+	NodeName string
+	Zone     string
+}
+
+type CollectorZonedNode struct {
+	ZonedNode
+	collector string
+}
+
+type TargetZonedNode struct {
+	ZonedNode
+	target string
+}
+
+func collectorZonedNodeKeyFunc(collectorZonedNode interface{}) (string, error) {
+	return collectorZonedNode.(CollectorZonedNode).collector, nil
+}
+
+func targetZonedNodeKeyFunc(targetZonedNode interface{}) (string, error) {
+	return targetZonedNode.(TargetZonedNode).target, nil
+}
+
+var _ Strategy = &perZoneStrategy{}
+
+type perZoneStrategy struct {
+	k8sClient              kubernetes.Interface
+	config                 consistent.Config
+	collectorToZonedNode   cache.Store
+	targetToZonedNode      cache.Store
+	consistentHasherByZone map[string]*consistent.Consistent
+}
+
+func newPerZoneStrategy(k8sClient kubernetes.Interface) Strategy {
+	config := consistent.Config{
+		PartitionCount:    1061,
+		ReplicationFactor: 5,
+		Load:              1.1,
+		Hasher:            hasher{},
+	}
+	perZoneStrategy := &perZoneStrategy{
+		k8sClient:              k8sClient,
+		config:                 config,
+		consistentHasherByZone: make(map[string]*consistent.Consistent),
+		collectorToZonedNode:   cache.NewTTLStore(collectorZonedNodeKeyFunc, cacheTTL),
+		targetToZonedNode:      cache.NewTTLStore(targetZonedNodeKeyFunc, cacheTTL),
+	}
+	return perZoneStrategy
+}
+
+func (s *perZoneStrategy) GetCollectorForTarget(collectors map[string]*Collector, item *target.Item) (*Collector, error) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	targetUrl := item.TargetURL
+	targetNodeName := item.GetNodeName()
+	targetZonedNode, exist, err := s.targetToZonedNode.GetByKey(targetUrl)
+	if err != nil {
+		fmt.Printf("failed to retrieve target zoned node from cache: %s", err)
+	}
+	if !exist || targetZonedNode.(TargetZonedNode).ZonedNode.NodeName != item.GetNodeName() {
+		k8sNode, err := s.retrieveK8sNode(ctx, targetNodeName)
+		if err != nil {
+			fmt.Printf("err retrieving k8s node %s for target %s: %s\n", item.GetNodeName(), targetUrl, err)
+		}
+		targetNodeZone, exist := k8sNode.ObjectMeta.Labels[v1.LabelTopologyZone]
+		if !exist {
+			fmt.Printf("succeeded to find the target node %s in the cluster but it doesn't support zone awareness", targetNodeName)
+		}
+		targetZonedNode = TargetZonedNode{
+			ZonedNode: ZonedNode{
+				NodeName: targetNodeName,
+				Zone:     targetNodeZone,
+			},
+			target: targetUrl,
+		}
+		err = s.targetToZonedNode.Add(targetZonedNode)
+		if err != nil {
+			fmt.Printf("error adding a cache for the node and zone information that relates to target %s: %s", targetUrl, err)
+		}
+	}
+
+	zonedConsistentHasher, exist := s.consistentHasherByZone[targetZonedNode.(TargetZonedNode).ZonedNode.Zone]
+	if !exist {
+		return nil, fmt.Errorf("unknown Zone %s", targetZonedNode.(TargetZonedNode).ZonedNode.Zone)
```

Comment:
Best to quote user-provided identifiers to avoid ambiguity when error messages get aggregated.

```suggestion
		return nil, fmt.Errorf("unknown Zone %q", targetZonedNode.(TargetZonedNode).ZonedNode.Zone)
```